pycernan/avro/client.py

```diff
@@ -1,47 +1,137 @@
 """
     Base Avro client from which all other clients derive.
 """
+import contextlib
 import socket
 
 import pycernan.avro.config
 
 from abc import ABCMeta, abstractmethod
+from queue import Queue, Empty
 
 from pycernan.avro.exceptions import EmptyBatchException
 from pycernan.avro.serde import serialize
 
 
+DefunctConnection = None
+
+
+class EmptyPoolException(Exception):
+    pass
+
+
+class TCPConnectionPool(object):
+    """
+    Fork friendly TCP connection pool.
+
+    Adapted from: https://github.com/gevent/gevent/blob/master/examples/psycopg2_pool.py
+    """
+    def __init__(self, host, port, maxsize, connect_timeout, read_timeout):
+        self.host = host
+        self.port = port
+
+        assert maxsize > 0
+        self.size = 0
+        self.maxsize = maxsize
+
+        self.connect_timeout = connect_timeout
+        self.read_timeout = read_timeout
+        self.pool = None
+
+    def create_connection(self):
+        sock = socket.create_connection((self.host, self.port), timeout=self.connect_timeout)
+        sock.settimeout(self.read_timeout)
+        return sock
+
+    def get(self, _block=True):
+        if not self.pool:
+            # Lazy queue creation, ensuring we are fork() friendly.
```

Comment:
You could check that the current pid matches a pid you store in the constructor or alongside a connection